ICE with closures in insert_autoderef_mappings

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```rust
pub enum Option<T> {
    Some(T),
    None,
}

pub use Option::{Some, None};

#[lang = "fn_once"]
pub trait FnOnce<Args> {
    #[lang = "fn_once_output"]
    type Output;

    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
}

impl<T> Option<T> {
    pub fn map<R, F: FnOnce(T) -> R>(self, f: F) -> Option<R> {
        match self {
            Some(value) => Some(f(value)),
            None => None,
        }
    }
}
```

### Meta

- What version of Rust GCC were you using, git sha if possible. 92b137048583c2c22c585097119dba6a5fe06228

### Error output

```
crab1: internal compiler error: in insert_autoderef_mappings, at rust/typecheck/rust-typecheck-context.cc:342
```

<!--
Include a backtrace if available.
-->
<details><summary>Backtrace</summary>
  <p>
  
  ```
crab1: internal compiler error: in insert_autoderef_mappings, at rust/typecheck/rust-typecheck-context.cc:342
0x76846b Rust::Resolver::TypeCheckContext::insert_autoderef_mappings(unsigned int, std::vector<Rust::Resolver::Adjustment, std::allocator<Rust::Resolver::Adjustment> >&&)
        ../../gcc/rust/typecheck/rust-typecheck-context.cc:342
0xbaf838 Rust::Resolver::coercion_site(unsigned int, Rust::TyTy::TyWithLocation, Rust::TyTy::TyWithLocation, Location)
        ../../gcc/rust/typecheck/rust-type-util.cc:219
0xb73077 Rust::TyTy::TypeCheckCallExpr::visit(Rust::TyTy::ADTType&)
        ../../gcc/rust/typecheck/rust-tyty-call.cc:64
0xb9ec00 Rust::TyTy::TypeCheckCallExpr::go(Rust::TyTy::BaseType*, Rust::HIR::CallExpr&, Rust::TyTy::VariantDef&, Rust::Resolver::TypeCheckContext*)
        ../../gcc/rust/typecheck/rust-tyty-call.h:39
0xb9ec00 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::CallExpr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:226
0xb9b85e Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0xb9cfb5 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::MatchExpr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:1399
0xb9b85e Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0xb9c886 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:537
0xb9b85e Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0xba95b1 Rust::Resolver::TypeCheckImplItem::visit(Rust::HIR::Function&)
        ../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:311
0xba723b Rust::Resolver::TypeCheckImplItem::Resolve(Rust::HIR::ImplBlock*, Rust::HIR::ImplItem*, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >)
        ../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:180
0xb90868 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ImplBlock&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:341
0xb8b3ec Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:49
0xb8b3ec Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:35
0xb572eb Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
        ../../gcc/rust/typecheck/rust-hir-type-check.cc:37
0xa3b3e1 Rust::Session::compile_crate(char const*)
        ../../gcc/rust/rust-session-manager.cc:620
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
  ```
  
  </p>
</details>

Got the ICE fixed in 2355 but now it results in:

```
test.rs:17:63: error: expected ‘Option<R>’ got ‘Option<<placeholder:>>’
   17 |     pub fn map<R, F: FnOnce(T) -> R>(self, f: F) -> Option<R> {
      |                                                     ~~~~~~    ^
```

We are missing a case to setup the associated types here with the FnOnce Trait

--- a/gcc/rust/typecheck/rust-tyty-bounds.cc
+++ b/gcc/rust/typecheck/rust-tyty-bounds.cc
@@ -227,22 +227,26 @@ TypeCheckBase::get_predicate_from_bound (HIR::TypePath &type_path,
 	std::vector<std::unique_ptr<HIR::Type>> inputs;
 	inputs.push_back (std::unique_ptr<HIR::Type> (implicit_tuple));
 
+	// resolve the fn_once_output type which assumes there must be an output
+	// set
+	rust_assert (fn.has_return_type ());
+	TypeCheckType::Resolve (fn.get_return_type ().get ());
+
+	HIR::TraitItem *trait_item = mappings->lookup_trait_item_lang_item (
+	  Analysis::RustLangItem::ItemType::FN_ONCE_OUTPUT);
+
+	std::vector<HIR::GenericArgsBinding> bindings;
+	Location output_locus = fn.get_return_type ()->get_locus ();
+	HIR::GenericArgsBinding binding (Identifier (
+					   trait_item->trait_identifier ()),
+					 fn.get_return_type ()->clone_type (),
+					 output_locus);
+	bindings.push_back (std::move (binding));
+
 	args = HIR::GenericArgs ({} /* lifetimes */,
 				 std::move (inputs) /* type_args*/,
-				 {} /* binding_args*/, {} /* const_args */,
-				 final_seg->get_locus ());
-
-	// resolve the fn_once_output type
-	TyTy::BaseType *fn_once_output_ty
-	  = fn.has_return_type ()
-	      ? TypeCheckType::Resolve (fn.get_return_type ().get ())
-	      : TyTy::TupleType::get_unit_type (
-		final_seg->get_mappings ().get_hirid ());
-	context->insert_implicit_type (final_seg->get_mappings ().get_hirid (),
-				       fn_once_output_ty);
-
-	// setup the associated type.. ??
-	// fn_once_output_ty->debug ();
+				 std::move (bindings) /* binding_args*/,
+				 {} /* const_args */, final_seg->get_locus ());
       }
       break;
 

--- a/gcc/rust/util/rust-hir-map.cc
+++ b/gcc/rust/util/rust-hir-map.cc
@@ -1137,5 +1137,18 @@ Mappings::lookup_builtin_marker ()
   return builtinMarker;
 }
 
+HIR::TraitItem *
+Mappings::lookup_trait_item_lang_item (Analysis::RustLangItem::ItemType item)
+{
+  DefId trait_item_id = UNKNOWN_DEFID;
+  bool trait_item_lang_item_defined = lookup_lang_item (item, &trait_item_id);
+
+  // FIXME
+  // make this an error? what does rustc do when a lang item is not defined?
+  rust_assert (trait_item_lang_item_defined);
+
+  return lookup_trait_item_defid (trait_item_id);
+}
+
 } // namespace Analysis
 } // namespace Rust

--- a/gcc/rust/util/rust-hir-map.h
+++ b/gcc/rust/util/rust-hir-map.h
@@ -321,6 +321,9 @@ public:
 
   HIR::ImplBlock *lookup_builtin_marker ();
 
+  HIR::TraitItem *
+  lookup_trait_item_lang_item (Analysis::RustLangItem::ItemType item);
+
 private:
   Mappings ();
 

--- a/gcc/testsuite/rust/compile/issue-2105.rs
+++ b/gcc/testsuite/rust/compile/issue-2105.rs
@@ -1,0 +1,23 @@
+pub enum Option<T> {
+    Some(T),
+    None,
+}
+
+pub use Option::{None, Some};
+
+#[lang = "fn_once"]
+pub trait FnOnce<Args> {
+    #[lang = "fn_once_output"]
+    type Output;
+
+    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
+}
+
+impl<T> Option<T> {
+    pub fn map<R, F: FnOnce(T) -> R>(self, f: F) -> Option<R> {
+        match self {
+            Some(value) => Some(f(value)),
+            None => None,
+        }
+    }
+}

